test/cpp/src/702-dram-refresh.cc

```diff
@@ -0,0 +1,101 @@
+    offset += LOG2_BLOCK_SIZE;
+    champsim::address_slice channel_slice{champsim::dynamic_extent{champsim::data::bits{champsim::lg2(DRAM_CHANNELS) + offset}, champsim::data::bits{offset}}, 0};
+    offset += champsim::lg2(DRAM_CHANNELS);
+    champsim::address_slice bank_slice{champsim::dynamic_extent{champsim::data::bits{champsim::lg2(DRAM_BANKS) + offset}, champsim::data::bits{offset}}, packet_num % DRAM_BANKS};
+    offset += champsim::lg2(DRAM_BANKS);
+    champsim::address_slice column_slice{champsim::dynamic_extent{champsim::data::bits{champsim::lg2(DRAM_COLUMNS) + offset}, champsim::data::bits{offset}}, 1};
+    offset += champsim::lg2(DRAM_COLUMNS);
+    champsim::address_slice rank_slice{champsim::dynamic_extent{champsim::data::bits{champsim::lg2(DRAM_RANKS) + offset}, champsim::data::bits{offset}}, packet_num % DRAM_RANKS};
+    offset += champsim::lg2(DRAM_RANKS);
+    champsim::address_slice row_slice{champsim::dynamic_extent{champsim::data::bits{64}, champsim::data::bits{offset}}, packet_num % DRAM_ROWS};
+    r.address = champsim::address{champsim::splice(row_slice, rank_slice, column_slice, bank_slice, channel_slice, block_slice)};
+    r.v_address = champsim::address{};
+    r.instr_id = 0;
+    r.response_requested = false;
+
+    if(r.type == access_type::LOAD)
+        channel->add_rq(r);
+    else
+        channel->add_wq(r);
+}
+
+std::vector<bool> refresh_test(MEMORY_CONTROLLER* uut, champsim::channel* channel_uut, uint64_t refresh_cycles, uint64_t DRAM_CHANNELS, uint64_t DRAM_RANKS, uint64_t DRAM_BANKS, uint64_t DRAM_COLUMNS, uint64_t DRAM_ROWS, champsim::chrono::picoseconds tREF)
+{
+    //a refresh MUST be done. If this is violated, then DRAM spec is violated.
+    
+    //record the refresh status of each bank
+    std::vector<bool> bank_refreshed(DRAM_BANKS,false);
+
+    //advanced current time to first refresh cycle, or test will fail since that is the trigger
+    uut->current_time += tREF;
+    //num of refresh cycles to cover
+    //record whether each refresh cycle was respected
+    std::vector<bool> refresh_done;
+
+    //we will cover the first 40 refreshes
+    uint64_t refresh_cycle = 2;
+    uint64_t cycles = 0;
+    while (uut->current_time < champsim::chrono::clock::time_point{} + tREF*refresh_cycles)
+    {
+        //generate a random packet
+        generate_packet(channel_uut,(uint64_t)cycles, DRAM_CHANNELS, DRAM_RANKS, DRAM_BANKS, DRAM_COLUMNS, DRAM_ROWS);
+        //operate mem controller
+        uut->_operate();
+        cycles++;
+        //make sure that for every refresh cycle, each bank undergoes refresh at least once
+        std::vector<bool> bank_under_refresh;
+        std::transform(std::begin(uut->channels[0].bank_request),std::end(uut->channels[0].bank_request),std::back_inserter(bank_under_refresh),[](const auto& entry){return(entry.under_refresh);});
```

Comment:
The problem is two-fold for Ramulator.
First, we can't see inside Ramulator easily. The ChampSimPlugin I wrote gathers as much data as it can.
The second is that Ramulator doesn't actually track the row under refresh. This is done internally in the DRAM chips typically, so the simulation doesn't _need_ to know it.

What we could do it grab the REFRESHES_ISSUED stat from ChampSimPlugin, and at least guarantee that the expected number of refreshes happened within the period. Though this will involve a lot of #ifdef splicing to pull off.

Reply:
How does refreshing show externally? Is the bank stalled until the refresh happens, perhaps? Let's write tests to observe things like that.